### ISSUE TITLE
[BUG] [main] ERROR com.nvidia.spark.rapids.GpuOverrideUtil - Encountered an exception applying GPU overrides java.lang.IllegalStateException: the broadcast must be on the GPU too

### ISSUE THREAD

**Issue by rohitreddy1698** [NVIDIA/spark-rapids#9469]
**What is your question?**
I am getting the above error when Im trying to run a the following query with RAPIDS enabled.

```
spark-sql> WITH parts      AS (SELECT partition.local_create_dt, partition.client_name, record_count
         >                     FROM local.download_100b.files
         >                     ORDER BY hash(partition.local_create_dt, partition.client_name)
         >                     LIMIT 57)
         >                     SELECT download_100b.*,
         >                            input_file_name()                                                   AS _file_name,
         >                            row_number() OVER (PARTITION BY input_file_name() ORDER BY hash(*)) AS rno
         >                     FROM local.download_100b
         >                            INNER JOIN parts
         >                            ON parts.local_create_dt = download_100b.local_create_dt
         >                              AND parts.client_name = download_100b.client_name
         > ;
```

The query works fine with RAPIDS disabled. Also the subquery where Im doing a select on local.download_100b.files works fine with RAPIDS enabled.

Here is the entire stack trace for the error :

```
06:40:10.707 [main] ERROR com.nvidia.spark.rapids.GpuOverrideUtil - Encountered an exception applying GPU overrides java.lang.IllegalStateException: the broadcast must be on the GPU too
java.lang.IllegalStateException: the broadcast must be on the GPU too
	at com.nvidia.spark.rapids.shims.GpuBroadcastJoinMeta.verifyBuildSideWasReplaced(GpuBroadcastJoinMeta.scala:54) ~[spark323/:?]
	at org.apache.spark.sql.rapids.execution.GpuBroadcastHashJoinMeta.convertToGpu(GpuBroadcastHashJoinExec.scala:45) ~[spark323/:?]
	at org.apache.spark.sql.rapids.execution.GpuBroadcastHashJoinMeta.convertToGpu(GpuBroadcastHashJoinExec.scala:26) ~[spark323/:?]
	at com.nvidia.spark.rapids.SparkPlanMeta.convertIfNeeded(RapidsMeta.scala:761) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuProjectExecMeta.convertToGpu(basicPhysicalOperators.scala:50) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuProjectExecMeta.convertToGpu(basicPhysicalOperators.scala:41) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.SparkPlanMeta.convertIfNeeded(RapidsMeta.scala:761) ~[spark3xx-common/:?]
	at org.apache.spark.sql.rapids.execution.GpuShuffleMetaBase.convertToGpu(GpuShuffleExchangeExecBase.scala:117) ~[spark3xx-common/:?]
	at org.apache.spark.sql.rapids.execution.GpuShuffleMetaBase.convertToGpu(GpuShuffleExchangeExecBase.scala:43) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.SparkPlanMeta.convertIfNeeded(RapidsMeta.scala:761) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuSortMeta.convertToGpu(GpuSortExec.scala:64) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuSortMeta.convertToGpu(GpuSortExec.scala:44) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.SparkPlanMeta.convertIfNeeded(RapidsMeta.scala:761) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuBaseWindowExecMeta.convertToGpu(GpuWindowExec.scala:158) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuBaseWindowExecMeta.convertToGpu(GpuWindowExec.scala:48) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.SparkPlanMeta.convertIfNeeded(RapidsMeta.scala:761) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuProjectExecMeta.convertToGpu(basicPhysicalOperators.scala:50) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuProjectExecMeta.convertToGpu(basicPhysicalOperators.scala:41) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.SparkPlanMeta.convertIfNeeded(RapidsMeta.scala:761) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuOverrides$.com$nvidia$spark$rapids$GpuOverrides$$doConvertPlan(GpuOverrides.scala:4073) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuOverrides.applyOverrides(GpuOverrides.scala:4412) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuOverrides.$anonfun$applyWithContext$3(GpuOverrides.scala:4278) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuOverrides$.logDuration(GpuOverrides.scala:452) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuOverrides.$anonfun$applyWithContext$1(GpuOverrides.scala:4275) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuOverrideUtil$.$anonfun$tryOverride$1(GpuOverrides.scala:4241) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuOverrides.applyWithContext(GpuOverrides.scala:4295) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuQueryStagePrepOverrides.$anonfun$apply$1(GpuOverrides.scala:4258) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuOverrideUtil$.$anonfun$tryOverride$1(GpuOverrides.scala:4241) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuQueryStagePrepOverrides.apply(GpuOverrides.scala:4261) ~[spark3xx-common/:?]
	at com.nvidia.spark.rapids.GpuQueryStagePrepOverrides.apply(GpuOverrides.scala:4254) ~[spark3xx-common/:?]
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec$.$anonfun$applyPhysicalRules$2(AdaptiveSparkPlanExec.scala:779) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126) ~[scala-library-2.12.17.jar:?]
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122) ~[scala-library-2.12.17.jar:?]
	at scala.collection.immutable.List.foldLeft(List.scala:91) ~[scala-library-2.12.17.jar:?]
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec$.applyPhysicalRules(AdaptiveSparkPlanExec.scala:778) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$initialPlan$1(AdaptiveSparkPlanExec.scala:180) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.<init>(AdaptiveSparkPlanExec.scala:179) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.adaptive.InsertAdaptiveSparkPlan.applyInternal(InsertAdaptiveSparkPlan.scala:63) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.adaptive.InsertAdaptiveSparkPlan.apply(InsertAdaptiveSparkPlan.scala:43) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.adaptive.InsertAdaptiveSparkPlan.apply(InsertAdaptiveSparkPlan.scala:40) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution$.$anonfun$prepareForExecution$1(QueryExecution.scala:434) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126) ~[scala-library-2.12.17.jar:?]
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122) ~[scala-library-2.12.17.jar:?]
	at scala.collection.immutable.List.foldLeft(List.scala:91) ~[scala-library-2.12.17.jar:?]
	at org.apache.spark.sql.execution.QueryExecution$.prepareForExecution(QueryExecution.scala:433) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executedPlan$1(QueryExecution.scala:157) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111) ~[spark-catalyst_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$1(QueryExecution.scala:183) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution.executePhase(QueryExecution.scala:183) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:157) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:150) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution.simpleString(QueryExecution.scala:201) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution.org$apache$spark$sql$execution$QueryExecution$$explainString(QueryExecution.scala:246) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.QueryExecution.explainString(QueryExecution.scala:215) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:98) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64) ~[spark-sql_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.hive.thriftserver.SparkSQLDriver.run(SparkSQLDriver.scala:69) ~[spark-hive-thriftserver_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.processCmd(SparkSQLCLIDriver.scala:384) ~[spark-hive-thriftserver_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.$anonfun$processLine$1(SparkSQLCLIDriver.scala:504) ~[spark-hive-thriftserver_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.$anonfun$processLine$1$adapted(SparkSQLCLIDriver.scala:498) ~[spark-hive-thriftserver_2.12-3.2.3.jar:3.2.3]
	at scala.collection.Iterator.foreach(Iterator.scala:943) [scala-library-2.12.17.jar:?]
	at scala.collection.Iterator.foreach$(Iterator.scala:943) [scala-library-2.12.17.jar:?]
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431) [scala-library-2.12.17.jar:?]
	at scala.collection.IterableLike.foreach(IterableLike.scala:74) [scala-library-2.12.17.jar:?]
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73) [scala-library-2.12.17.jar:?]
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56) [scala-library-2.12.17.jar:?]
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.processLine(SparkSQLCLIDriver.scala:498) [spark-hive-thriftserver_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver$.main(SparkSQLCLIDriver.scala:287) [spark-hive-thriftserver_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.main(SparkSQLCLIDriver.scala) [spark-hive-thriftserver_2.12-3.2.3.jar:3.2.3]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52) [spark-core_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:955) [spark-core_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180) [spark-core_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203) [spark-core_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90) [spark-core_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1043) [spark-core_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1052) [spark-core_2.12-3.2.3.jar:3.2.3]
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala) [spark-core_2.12-3.2.3.jar:3.2.3]
```

I am unable to understand what spark is trying to place on the GPU.
I have setup a spark stand alone cluster with 1 leader and 3 worker nodes.

with spark-env.sh :
```
export SPARK_HOME=/root/spark-3.2.3-bin-hadoop3.2
export SPARK_RAPIDS_DIR=/root/rapids-jars/
export SPARK_RAPIDS_PLUGIN_JAR=${SPARK_RAPIDS_DIR}/rapids-4-spark_2.12-23.02.0.jar
export SPARK_MASTER_HOST=mr3-0509-0116-10-jea.mr3.simcloud.apple.com
export SPARK_MASTER_PORT=7077
export SPARK_MASTER_WEBUI_PORT=8080
export SPARK_WORKER_PORT=8000
export SPARK_WORKER_WEBUI_PORT=8081
export SPARK_WORKER_CORES=24
export SPARK_WORKER_MEMORY=60g
export SPARK_WORKER_OPTS="-Dspark.worker.resource.gpu.amount=2 -Dspark.worker.resource.gpu.discoveryScript=/root/getGpusResources.sh"
```

and spark-defaults.conf :
```
spark.executor.memory 16g
spark.executor.cores 8
spark.executor.resource.gpu.amount 1
spark.rapids.sql.concurrentGpuTasks 4
spark.executor.resource.gpu.discoveryScript /root/getGpusResources.sh
spark.sql.files.maxPartitionBytes 1g
spark.rapids.memory.pinnedPool.size 25g

spark.sql.adaptive.coalescePartitions.enabled true
spark.sql.adaptive.join.enabled true
spark.sql.autoBroadcastJoinThreshold 50MB
spark.sql.adaptive.skewJoin.enabled true

spark.task.resource.gpu.amount 0.15
spark.rapids.sql.enabled true
```

I have 2 GPUs per node each with **16384MiB** of memory.
I have 3 worker nodes , on which 6 executors are getting started
So now each executor is getting 1 GPU , and is using a maximum of 
0.15 x 4 = 0.6 of the GPU leaving 0.4 GPU still available why is it still unable to broadcast.

I am also unable to see the plan on the UI. I am unable to understand what is being broadcasted.

Would installing the profiling tool help ?

**Comment by revans2** [comment]
@rohitreddy1698 This looks like a bug in our code. We will try and reproduce this and get back to you on how we will try and fix this.

**Comment by revans2** [comment]
> Would installing the profiling tool help ?

No this is crashing before the SQL finishes compiling into an RDD. So it would not show up in the history log at all.

**Comment by revans2** [comment]
@rohitreddy1698 what version of the plugin are you running with? I'll try a few released versions to see if I can find it, but I could not reproduce the error on Spark 3.3.0 with our latest development branch. I'll switch to Spark 3.2.3 just to match you too.

**Comment by revans2** [comment]
Also @rohitreddy1698 what version of iceberg are you running with? I tried both 1.0.0 and 1.4.0 and neither of them produce the metadata schema that you are querying. I'll keep looking but more info would really help.

**Comment by revans2** [comment]
Okay I have reproduced the error. I ended up getting confused because I am not an iceberg expert and didn't realize that we had to partition the data by local_create_dt and client_name. I'll try to dig in and see what is happening, but I am going to transform this into a bug instead of a question now.

**Comment by revans2** [comment]
Here are my instructions on how to reproduce the error.

My command to run spark.
```
SPARK_CONF_DIR=spark_conf/gpu/ spark_3.2.3/bin/spark-shell --jars datagen/target/datagen_2.12-23.12.0-SNAPSHOT-spark323.jar --packages org.apache.iceberg:iceberg-spark-runtime-3.2_2.12:1.0.0 --conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions --conf spark.sql.catalog.spark_catalog=org.apache.iceberg.spark.SparkSessionCatalog --conf spark.sql.catalog.spark_catalog.type=hive --conf spark.sql.catalog.local=org.apache.iceberg.spark.SparkCatalog --conf spark.sql.catalog.local.type=hadoop --conf spark.sql.catalog.local.warehouse=/data/tmp/warehouse/
```

The query to generate the data

```
import org.apache.spark.sql.tests.datagen._

val dbgen = new DBGen()
dbgen.addTable("download_100b", """local_create_dt string, client_name string, other long""", 1000000L)
dbgen("download_100b")("local_create_dt").setSeedRange(0, 10)
dbgen("download_100b")("client_name").setSeedRange(0, 10)
val df = dbgen("download_100b").toDF(spark).repartition(col("local_create_dt"), col("client_name"))

spark.sql("""CREATE TABLE local.download_100b (local_create_dt string, client_name string, other long) using iceberg partitioned by (local_create_dt, client_name)""")
df.writeTo("local.download_100b").append()
```

Then run the original query itself.

```
val QUERY = """WITH parts      AS (SELECT partition.local_create_dt, partition.client_name, record_count
                               FROM local.download_100b.files
                               ORDER BY hash(partition.local_create_dt, partition.client_name)
                               LIMIT 57)
                               SELECT download_100b.*,
                                      input_file_name()                                                   AS _file_name,
                                      row_number() OVER (PARTITION BY input_file_name() ORDER BY hash(*)) AS rno
                               FROM local.download_100b
                                      INNER JOIN parts
                                      ON parts.local_create_dt = download_100b.local_create_dt
                                        AND parts.client_name = download_100b.client_name"""

spark.sql(QUERY).show(100, false)
```

This results in the error.

**Comment by revans2** [comment]
@rohitreddy1698 I think you need to disable iceberg support.

```
spark.conf.set("spark.rapids.sql.format.iceberg.enabled", false)
```

TLDR;

The bug you are seeing appears to be related to another bug I filed when first trying to debug this one.

https://github.com/NVIDIA/spark-rapids/issues/9473

Because the check walks through the broadcast it ends up putting part of the plan on the CPU when there was no reason to do so. But the other half of the join ended up not being on the CPU so it failed with the error above.

But when I "fixed" it, it failed because we don't support iceberg 1.0.0.  When I disable iceberg all of it just works after that, but the first half of the processing is all on the CPU.

**Comment by rohitreddy1698** [comment]
Hi , Thanks a lot for your prompt response ! I was not expecting a reply so soon.
I am sorry I should've mentioned the versions in the initial question itself.

These were the versions I was using :
SPARK_VERSION=${SPARK_VERSION:-3.2.3}
HADOOP_VERSION=${HADOOP_VERSION:-3.3.4}
ORC_VERSION=1.7.3
ICEBERG_VERSION=0.13.2
LOG4J2_VERSION=2.17.1
AWS_SDK2_VERSION=2.17.135
SCALA_VERSION=2.12.17

The Iceberg we are using 0.13.2 with some changes over it internal to the company.

No this is crashing before the SQL finishes compiling into an RDD. So it would not show up in the history log at all. - Got it Thank you.

The RAPIDS version : rapids-4-spark_2.12-23.02.0

Also Thanks a lot for figuring out so much info , I just realised how much info was missing in my original question after reading your responses. Will try to include all this. 

Will go through the link to the original bug you have provided. and will try disabling the iceberg support to see if this works.



Also I want to know a little bit more about the bug, can I reach out to you on slack?

Regards,
Rohit Mothe

**Comment by rohitreddy1698** [comment]
Hi ,

I was able to get over the issue by disabling the iceberg support , but want to understand more about the issue.

Regards,
Rohit Mothe

**Comment by revans2** [comment]
There are a few issues here, and incomplete workarounds happening that are causing the problem.

The original problem is with trying to get the name of the file being read.  Spark does this as an out of bound operation. It sets a thread local variable that is read later on by an expression. This works okay for row at a time processing, but when we start to process the data in batches there are new corner cases which can cause problems. Like in this case when there is CPU processing in between reading the file and the name of the file is put into the dataframe. To work around this we walk the query plan and put some things back into the CPU so that the file name shows up in the correct place.  But that workaround had some bugs in it. It didn't expect to see a join in between the file read and the file name being added to the dataframe. So it didn't handle any join cases correctly. This is where things are getting messed up. It ended up putting the join on the GPU but thought it needed to put the broadcast on the CPU, which it didn't need to do. It just got confused by the join.
